File: plugins/markdown/src/org/intellij/plugins/markdown/lang/psi/MarkdownASTFactory.java
package org.intellij.plugins.markdown.lang.psi;

import com.intellij.lang.ASTFactory;
import com.intellij.psi.impl.source.tree.CompositeElement;
import com.intellij.psi.impl.source.tree.LeafElement;
import com.intellij.psi.tree.IElementType;
import org.intellij.plugins.markdown.lang.MarkdownElementTypes;
import org.intellij.plugins.markdown.lang.MarkdownTokenTypes;
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownCodeFenceContentImpl;
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownCodeFenceImpl;
import org.intellij.plugins.markdown.lang.psi.impl.MarkdownTableSeparatorRow;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class MarkdownASTFactory extends ASTFactory {
  @Nullable
  @Override
  public CompositeElement createComposite(@NotNull IElementType type) {
    if (type == MarkdownElementTypes.CODE_FENCE) {
      return new MarkdownCodeFenceImpl(type);
    }

    return super.createComposite(type);
  }

  @Nullable
  @Override
  public LeafElement createLeaf(@NotNull IElementType type, @NotNull CharSequence text) {
    if (type == MarkdownTokenTypes.CODE_FENCE_CONTENT) {
      return new MarkdownCodeFenceContentImpl(type, text);
    }
    if (type == MarkdownTokenTypes.TABLE_SEPARATOR && text.length() > 1) {
      return new MarkdownTableSeparatorRow(text);
    }
    return super.createLeaf(type, text);
  }
}

File: plugins/markdown/core/src/org/intellij/plugins/markdown/lang/psi/MarkdownAstFactory.kt
// Copyright 2000-2021 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.intellij.plugins.markdown.lang.psi

import com.intellij.lang.ASTFactory
import com.intellij.psi.impl.source.tree.CompositeElement
import com.intellij.psi.impl.source.tree.LeafElement
import com.intellij.psi.tree.IElementType
import org.intellij.plugins.markdown.lang.MarkdownElementTypes
import org.intellij.plugins.markdown.lang.MarkdownTokenTypeSets
import org.intellij.plugins.markdown.lang.MarkdownTokenTypes
import org.intellij.plugins.markdown.lang.psi.impl.*
import org.jetbrains.annotations.ApiStatus

@ApiStatus.Internal
class MarkdownAstFactory: ASTFactory() {
  override fun createComposite(type: IElementType): CompositeElement? {
    return when (type) {
      MarkdownElementTypes.CODE_FENCE -> MarkdownCodeFence(type)
      MarkdownElementTypes.FRONT_MATTER_HEADER -> MarkdownFrontMatterHeader(type)
      else -> super.createComposite(type)
    }
  }

  override fun createLeaf(type: IElementType, text: CharSequence): LeafElement? {
    return when {
      type == MarkdownTokenTypes.LIST_NUMBER -> MarkdownListNumber(type, text)
      type == MarkdownTokenTypes.CODE_FENCE_CONTENT -> MarkdownCodeFenceContent(type, text)
      type == MarkdownElementTypes.FRONT_MATTER_HEADER_CONTENT -> MarkdownFrontMatterHeaderContent(type, text)
      type == MarkdownTokenTypes.TABLE_SEPARATOR && text.length > 1 -> MarkdownTableSeparatorRow(text)
      type == MarkdownTokenTypes.TABLE_SEPARATOR && text.toString() == "|" -> MarkdownTableSeparator(text)
      type == MarkdownElementTypes.COMMENT_VALUE -> MarkdownCommentValue(text)
      type in MarkdownTokenTypeSets.AUTO_LINKS -> MarkdownAutoLink(type, text)
      else -> super.createLeaf(type, text)
    }
  }
}
